runtime/src/main/java/org/corfudb/runtime/GarbageInformer.java

```diff
@@ -148,35 +155,37 @@ public void add(long markerAddress, List<SMRRecordLocator> locators) {
         return new ArrayList<>(garbage.values());
     }
 
+    public void submitGCTask() {
+        try {
+            drainExecutor.execute(this::gcUnsafe);
+        } catch (RejectedExecutionException ex) {
+            log.trace("drain executor reject execution");
+        }
+    }
+
     /**
      * Drains garbage decisions from receiving queue and sends them to LogUnit servers.
      */
-    public synchronized void gc() {
-        // drains sending queue first when it reaches the capacity limit.
-        if (garbageSendingDeque.size() >= SENDING_QUEUE_CAPACITY) {
-            log.debug("GarbageInformer: Drains sending queue");
-            sendGarbage();
-        }
-
-        // If garbageSendingDeque is still full, gives up drain garbageReceivingQueue.
-        if (garbageSendingDeque.size() >= SENDING_QUEUE_CAPACITY) {
-            log.debug("GarbageInformer: Stop drains receiving queue because sending queue is full");
-            return;
+    @VisibleForTesting
+    public void gcUnsafe() {
+        while (garbageSendingDeque.size() < SENDING_QUEUE_CAPACITY) {
+            List<SMRGarbageEntry> garbageEntries = new ArrayList<>();
+            int drainedNum = garbageReceivingQueue.drainTo(garbageEntries, BATCH_SIZE);
```

Comment:
Check drainedNum == 0 right after this line? Otherwise your batch could be zero.

Reply:
done.